fgpyo/fasta/builder.py

```diff
@@ -0,0 +1,231 @@
+
+        Example
+        add("AAA", 2) results in the following bases -> "AAAAAA"
+        """
+        # Remove any spaces in string and enforce upper case format
+        bases = bases.replace(" ", "").upper()
+        self.bases += str(bases * times)
+        return self
+
+
+class FastaBuilder:
+    """Builder for constructing sets of one or more contigs.
+
+    Provides the ability to manufacture sets of contigs from minimal input, and automatically
+    generates the information necessary for writing the FASAT file, index, and dictionary.
+
+    A builder is constructed from an assembly, species, and line length. All attributes have
+    defaults, however these can be overwritten.
+
+    Contigs are added to FastaBuilder using:
+    :func:`~fgpyo.fasta.builder.FastaBuilder.add`
+
+    Bases are added to existing contigs using:
+    :func:`~fgpyo.fasta.builder.FastaBuilder.add.add`
+
+    Once accumulated the contigs can be written to a file using:
+    :func:`~fgpyo.fasta.builder.FastaBuilder.to_file`
+
+    Calling to_file() will also generate the fasta index (.fai) and sequence dictionary (.dict).
+
+    Attributes:
+        assembly: Assembly information, if None default is 'testassembly'
+        species: Species, if None default is 'testspecies'
+        line_length: Desired line length, if None default is 80
+        contig_builders: Private dictionary of contig names and instances of ContigBuilder
+    """
+
+    def __init__(
+        self,
+        assembly: str = "testassembly",
+        species: str = "testspecies",
+        line_length: int = 80,
+    ):
+        self.assembly: str = assembly
+        self.species: str = species
+        self.line_length: int = line_length
+        self.__contig_builders: Dict[str, ContigBuilder] = {}
+
+    def __getitem__(self, key: str) -> ContigBuilder:
+        """ Access instance of ContigBuilder by name """
+        return self.__contig_builders[key]
+
+    def add(
+        self,
+        name: str,
+        assembly: str = "testassembly",
+        species: str = "testspecies",
+    ) -> ContigBuilder:
+        """
+        Creates and returns a new ContigBuilder for a contig with the provided name.
+        Contig names must be unique, attempting to create two seperate contigs with the same
+        name will result in an error.
+
+        Args:
+            name: Unique contig ID, ie., "chr10"
+            assembly: Assembly information, if None default is 'testassembly'
+            species: Species information, if None default is 'testspecies'
+        """
+        # Check if name has already been used
+        # If name already exists raise exception
+        # Else create instance of ContigBuilder and add to self.__contig_builders
+        if name in self.__contig_builders:
+            raise Exception(
+                f"The contig {name} already exists, see docstring for methods on "
+                f"adding bases to existing contigs"
+            )
```

Comment:
Can we get rid of the if/else altogether.
```suggestion
        assert name not in self.__contig_builders, (
            f"The contig {name} already exists, see docstring for methods on "
            f"adding bases to existing contigs"
        )
```